monitoring/aws/lambda/ci_monitor_lambda_function.py

```diff
@@ -0,0 +1,116 @@
+#!/usr/bin/env python3
+
+import boto3
+import datetime
+import json
+import io
+
+NOW = datetime.datetime.utcnow()
+
+# Our CI job is configured to use cluster names of the form:
+# hiveci-$timestamp-$pull
+NAME_FILTER = dict(Name='tag:Name', Values=['hiveci-*'])
+# Where $timestamp is the number of seconds since the epoch, in hex; and
+# $pull is the PR number, also in hex.
+def parse_vpc(vpc):
+    ret = dict()
+    ret['id'] = vpc['VpcId']
+    vpc_name = get_tag_value(vpc, key='Name')
+    _, hex_stamp, hex_pr, _ = vpc_name.split('-', 3)
+    ret['datetime'] = datetime.datetime.utcfromtimestamp(int(hex_stamp, 16))
+    ret['age'] = NOW - ret['datetime']
+    ret['pr'] = int(hex_pr, 16)
+    return ret
+
+
+def get_tag_value(o, key=None):
+    for tag in o.get('Tags', []):
+        if tag['Key'] == key:
+            return tag['Value']
+    return None
+
+
+def lambda_handler(event, context):
+
+    print(event)
+    debug = bool(event.get('debug'))
+    vpcs = {}
+
+    for region in event["regions"]:
+        ec2client = boto3.client('ec2', region)
+
+        try:
+            response = ec2client.describe_vpcs(Filters=[NAME_FILTER])
+        except Exception as e:
+            print("Error fetching VPCs from region %s: %s" % (region, e))
```

Comment:
Can we save the error so that it can be included in the periodic email so we aren't silently not watching for VPCs in a region?

Reply:
Good call. (Could make this change to the "running instances" bot too.

Reply:
Hmph. The problem I see here is: in my testing, a couple of regions were consistently busted. If that's a consistent issue, and we send the email any time it's nonempty, we'll get an email every 4h. That doesn't seem sustainable.

Open to ideas here. The CI job *seems* to always create our clusters in us-east-1. If we can confirm that, I can trim the whole thing down to just check that region. Then I can reliably send the email if a) we get hits, or b) the fetch fails.

Reply:
Maybe Thursday was just a bad day. The regions all seem to be okay now. I've updated the code accordingly.